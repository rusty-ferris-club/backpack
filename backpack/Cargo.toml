[package]
name = "backpack"
version = "2.1.1"
edition = "2021"


#
# == This is a dual CLI + library crate. ==
#
# * To turn this into a pure library crate, remove [[features]] and [[bin]],
#   and remove optional deps.
# * To turn into a pure CLI crate, do the same as before, but don't remove
#   optional deps, turn them into non-optional.
#
# To use both as dual crate:
#
# 1. For a CLI, nothing needs to be done.
# 2. For a library, build using `--no-default-features`, and depend using `default-features = false`
#


[dependencies]

clap = { version = "3", features = ["cargo"], optional = true }
serde = "1"
serde_json = "1"
serde_derive = "1"
serde_yaml = "^0.8.24"
serde_regex = "^1.1.0"
env_logger = "^0.9.0"
log = "^0.4.17"
anyhow = "1"
console = "^0.15.0"
url = "2"
git-url-parse = "^0.4.2"
lazy_static = "1"
regex = "1"
cached-path = "^0.5.3"
tempfile = "3"
walkdir = "2"
dirs = "4"
tracing = "^0.1.34"
merge-struct = "0.1.0"
reqwest = { version = "^0.11.11", features = ["blocking"] }
requestty = "0.4.0"
requestty-ui = "0.4.0"
tracing-forest = { version = "^0.1.4" }
tracing-subscriber = { version = "^0.3.11", features = ["env-filter"] }
# interactive-actions = "1.0.1"
# interactive-actions =  { path = "../../interactive-actions/interactive-actions" }
interactive-actions =  { git = "https://github.com/strowk/interactive-actions.git" }
tera = "^1.17.0"
tera-text-filters = "^1.0.0"
content_inspector = "0.2.4"
edit = "*"

[dev-dependencies]
insta = { version = "1.17.1", features = ["backtrace", "redactions"] }
fs_extra = "^1.2.0"
rstest = "^0.14.0"
pretty_assertions = "1"
serial_test = "0.4.0"


[features]
default = ["cli"]
# list optionals here:
cli = ["clap"]

[[bin]]
name = "bp"
required-features = ["cli"]
