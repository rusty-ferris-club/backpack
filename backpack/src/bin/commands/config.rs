use anyhow::Result as AnyResult;
use backpack::config::Config;
use clap::{Arg, ArgMatches, Command};
use std::path::Path;

pub fn command() -> Command<'static> {
    Command::new("config")
        .about("Create a personal configuration")
        .arg(
            Arg::new("init")
                .short('i')
                .long("init")
                .help("Initialize an empty configuration file")
                .takes_value(false),
        )
        .arg(
            Arg::new("local")
                .long("local")
                .help("Initialize local configuration")
                .takes_value(false),
        )
}

fn print_path(kind: &str, path: &Path) {
    println!(
        "{} {} {}",
        kind,
        if path.exists() { "(found):" } else { "(none):" },
        path.display()
    );
}
pub fn run(_matches: &ArgMatches, subcommand_matches: &ArgMatches) -> AnyResult<bool> {
    if subcommand_matches.is_present("init") {
        let generated = if subcommand_matches.is_present("local") {
            Config::init_local()?
        } else {
            Config::init_global()?
        };
        println!("wrote: {}.", generated.display());
    } else {
        let local = Config::local_config_file();
        let global = Config::global_config_file()?;
        print_path("global", global.as_path());
        print_path("local", local.as_path());

        let t = Config::load_or_default()?.0.to_text()?;
        println!("{}", t);
    }

    Ok(true)
}
