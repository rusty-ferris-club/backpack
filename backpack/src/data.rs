use anyhow::Result as AnyResult;
use core::fmt::Debug;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Clone, Debug)]
pub struct Location {
    pub url: String,
    pub domain: String,
    pub path: String,
    pub project: String,
    pub port: Option<u16>,
    pub query: Option<String>,
    pub subfolder: Option<String>,
    pub gref: Option<String>,
    pub username: String,
    pub password: Option<String>,
    pub is_git: bool,
}
impl Location {
    pub fn git_url(&self) -> String {
        format!(
            "git@{}:{}.git",
            self.domain,
            self.path.trim_start_matches('/')
        ) // needs: auth, port
    }

    pub fn web_url(&self) -> String {
        format!("https://{}{}", self.domain, self.path) // needs: auth, port
    }

    /// Create a logical `Location` from a `Url`.
    ///
    /// # Errors
    ///
    /// This function will return an error if `url` is illegal
    pub fn from(url: &Url, is_git: bool) -> AnyResult<Self> {
        let path = url.path();
        let parts = path.split("/-/").collect::<Vec<_>>();
        let (path, subfolder) = if parts.len() == 2 {
            (parts[0], Some(parts[1].to_string()))
        } else {
            (path, None)
        };

        let project = path
            .split('/')
            .last()
            .ok_or_else(|| anyhow::anyhow!("cannot find project name"))?;

        Ok(Self {
            url: url.to_string(),
            domain: url.domain().unwrap_or_default().to_string(),
            path: path.trim_end_matches(".git").to_string(),
            project: project.to_string(),
            port: url.port(),
            query: url.query().map(str::to_string),
            subfolder,
            gref: url.fragment().map(str::to_string),
            username: url.username().to_string(),
            password: url.password().map(str::to_string),
            is_git,
        })
    }
}

#[derive(Debug)]
pub struct Archive {
    pub url: String,
    pub root: ArchiveRoot,
}

#[derive(Debug)]
pub enum ArchiveRoot {
    Folder(String),
    FirstFolder,
    None,
}
#[derive(Debug)]
pub struct Assets {
    pub archive: Option<Archive>,
    pub git: Option<String>,
}

pub enum Overwrite {
    Ask,
    Always,
    Never,
    Custom(Box<dyn Fn(&str) -> bool>),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum CopyMode {
    #[serde(rename = "new")]
    Copy,
    #[serde(rename = "apply")]
    Apply,
    #[serde(rename = "all")]
    All,
}

#[allow(clippy::struct_excessive_bools)]
#[derive(Debug)]
pub struct Opts {
    pub show_progress: bool,
    pub overwrite: bool,
    pub mode: CopyMode,
    pub is_git: bool,
    pub no_cache: bool,
    pub always_yes: bool,
    pub remote: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_location_urls() {
        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            true,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc =
            Location::from(&Url::parse("https://github.com/user/repo").unwrap(), false).unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            false,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");
    }
}
