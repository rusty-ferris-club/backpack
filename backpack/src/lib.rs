//#![deny(clippy::pedantic)]
//#![deny(clippy::nursery)]
#![allow(clippy::must_use_candidate)]
#![allow(clippy::module_name_repetitions)]
#![allow(clippy::use_self)]
#![allow(clippy::unused_self)]
#![feature(path_try_exists)]

pub mod config;
pub mod content;
pub mod data;
pub mod fetch;
pub mod git;
mod merge;
pub mod prompt;
pub mod run;
pub mod shortlink;
pub mod vendors;
