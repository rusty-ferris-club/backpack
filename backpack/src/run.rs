use crate::config::Config;
use crate::content::Deployer;
use crate::data::CopyMode;
use crate::fetch::Fetcher;
use crate::git::{GitCmd, GitProvider};
use crate::prompt::Prompt;
use crate::shortlink::Shortlink;
use anyhow::{Context, Result};
use std::path::Path;

#[allow(clippy::struct_excessive_bools)]
#[derive(Debug)]
pub struct Opts {
    pub overwrite: bool,
    pub mode: CopyMode,
    pub is_git: bool,
    pub no_cache: bool,
    pub no_dest_input: bool,
    pub always_yes: bool,
    pub remote: Option<String>,
}
pub struct Runner {
    git: Box<dyn GitProvider>,
    pub show_progress: bool,
}

impl Default for Runner {
    fn default() -> Self {
        Self {
            git: Box::new(GitCmd::default()),
            show_progress: false,
        }
    }
}
impl Runner {
    /// Run the workflow with progress
    ///
    /// # Errors
    ///
    /// This function will return an error if anything in the workflow failed
    pub fn run(&self, shortlink: Option<&str>, dest: Option<&str>, opts: &Opts) -> Result<()> {
        self.run_workflow(shortlink, dest, opts)
    }

    #[tracing::instrument(skip(self), err)]
    fn run_workflow(&self, shortlink: Option<&str>, dest: Option<&str>, opts: &Opts) -> Result<()> {
        let (mut config, _) = Config::load_or_default().context("could not load configuration")?;

        // optionally add remote and sync here if remote exists
        if let Some(remote) = opts.remote.as_ref() {
            let num = config.fetch_and_load_remote_projects(remote.as_str())?;
            let prompt = Prompt::new(&config);
            if prompt.confirm_save_remotes(num)? {
                config.save()?;
            }
        }

        let config = config;
        let prompt = Prompt::new(&config);
        let should_confirm = shortlink.is_none() || dest.is_none();

        let (is_git, shortlink, dest) = prompt.fill_missing(shortlink, dest, opts)?;

        // confirm
        if !opts.always_yes
            && should_confirm
            && !prompt.are_you_sure(&shortlink, dest.as_deref())?
        {
            return Ok(());
        }

        if self.show_progress {
            prompt.say_resolving();
        }

        let sl = Shortlink::new(&config, self.git.as_ref());
        let (location, assets) = sl.resolve(&shortlink, is_git)?;

        let cached_path = Config::global_cache_folder()?;
        let fetcher = Fetcher::new(self.git.as_ref(), cached_path.as_path());
        if self.show_progress {
            prompt.say_fetching();
        }
        let (source, remove_source) = fetcher.fetch(&location, &assets, opts.no_cache)?;

        if self.show_progress {
            prompt.say_unpacking();
        }
        let deployer = Deployer::default();
        let res = deployer.deploy(
            Path::new(&source),
            dest.as_ref().map(Path::new),
            &location,
            &opts.mode,
            opts.overwrite,
            remove_source,
        )?;
        if self.show_progress {
            prompt.say_done(&res);
        }
        Ok(())
    }
}
