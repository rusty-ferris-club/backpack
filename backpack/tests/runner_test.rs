use std::{env, fs};

use anyhow::Result;
use backpack::config::Config;
use backpack::data::{CopyMode, Opts};
use backpack::run::{Runner, RunnerEvents};
use insta::{assert_debug_snapshot, assert_yaml_snapshot};
use requestty_ui::events::KeyCode;
use serial_test::serial;
use walkdir::{DirEntry, WalkDir};

fn ensure_no_config() {
    env::set_var("BP_FOLDER", ".backpack-none");
    env::set_var("BP_CONF", "none.yaml");

    let global_folder = Config::global_config_folder().unwrap();
    assert!(global_folder.ends_with(".backpack-none"));
    if global_folder.exists() {
        fs::remove_dir_all(&global_folder).unwrap();
    }
}

fn list_folder(dest: &str) -> Vec<String> {
    let mut r = WalkDir::new(dest)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|file| file.metadata().unwrap().is_file())
        .map(DirEntry::into_path)
        .map(|p| p.to_string_lossy().replace('\\', "/"))
        .collect::<Vec<_>>();
    r.sort();
    r
}

fn run(
    shortlink: Option<&str>,
    dest: Option<&str>,
    mode: CopyMode,
    local_config: Option<&str>,
    is_git: bool,
    events: Option<RunnerEvents>,
) -> Result<Vec<String>> {
    let tests_out = "tests-out";
    fs::remove_dir_all(tests_out).ok();
    fs::create_dir(tests_out).unwrap();
    let current_dir = env::current_dir().unwrap();
    env::set_current_dir(tests_out).unwrap();
    if let Some(config) = local_config {
        let _res = fs::remove_file(".backpack.yaml");
        fs::write(".backpack.yaml", config).unwrap();
    }

    if let Some(events) = events {
        Runner::default().run_with_events(
            shortlink,
            dest,
            &Opts {
                show_progress: false,
                overwrite: false,
                is_git,
                no_cache: false,
                always_yes: true,
                remote: None,
                config_file: None,
                mode,
            },
            &events,
        )?;
    } else {
        Runner::default().run(
            shortlink,
            dest,
            &Opts {
                show_progress: false,
                overwrite: false,
                is_git: false,
                no_cache: false,
                always_yes: true,
                remote: None,
                config_file: None,
                mode,
            },
        )?;
    };

    env::set_current_dir(current_dir).unwrap();
    Ok(list_folder(tests_out))
}

fn run_with_no_config(
    shortlink: Option<&str>,
    dest: Option<&str>,
    mode: CopyMode,
    is_git: bool,
    events: Option<RunnerEvents>,
) -> Result<Vec<String>> {
    ensure_no_config();
    run(shortlink, dest, mode, None, is_git, events)
}

fn run_with_local_config(
    shortlink: Option<&str>,
    dest: Option<&str>,
    mode: CopyMode,
    config: &str,
    is_git: bool,
    events: Option<RunnerEvents>,
) -> Result<Vec<String>> {
    ensure_no_config();
    env::remove_var("BP_CONF");
    run(shortlink, dest, mode, Some(config), is_git, events)
}

#[test]
#[serial]
fn test_run_source_dest() {
    assert_debug_snapshot!(run_with_no_config(
        Some("rusty-ferris-club/backpack-e2e-frozen"),
        Some("out"),
        CopyMode::Copy,
        false,
        None,
    ));
}

#[test]
#[serial]
fn test_run_source_dest_subfolder() {
    assert_debug_snapshot!(run_with_no_config(
        Some("rusty-ferris-club/backpack-e2e-frozen/-/.github"),
        Some("out"),
        CopyMode::Copy,
        false,
        None,
    ));
}

#[test]
#[serial]
fn test_run_source_dest_single_file() {
    assert_debug_snapshot!(run_with_no_config(
        Some("rusty-ferris-club/backpack-e2e-frozen/-/.github/workflows/build.yml"),
        Some("out/build.yml"),
        CopyMode::Copy,
        false,
        None,
    ));
}

#[test]
#[serial]
fn test_run_source_single_file() {
    assert_debug_snapshot!(run_with_no_config(
        Some("rusty-ferris-club/backpack-e2e-frozen/-/.github/workflows/build.yml"),
        None,
        CopyMode::Apply,
        false,
        Some(RunnerEvents {
            prompt_events: Some(vec![
                KeyCode::Enter.into(),     // no dest
            ]),
            actions_events: None,
        }),
    ));
}

#[test]
#[serial]
fn test_run_source_gist() {
    assert_debug_snapshot!(run_with_no_config(
        Some("https://gist.github.com/jondot/15086f59dab44f30bb10f82ca09f4887"),
        None,
        CopyMode::Apply,
        false,
        Some(RunnerEvents {
            prompt_events: Some(vec![
                KeyCode::Enter.into(),     // no dest
            ]),
            actions_events: None,
        }),
    ));
}

#[test]
#[serial]
fn test_run_with_local_project_actions() {
    assert_yaml_snapshot!(run_with_no_config(
        Some("rusty-ferris-club/backpack-e2e-frozen-localproj"),
        None,
        CopyMode::Copy,
        false,
        Some(RunnerEvents {
            prompt_events: Some(vec![
                KeyCode::Enter.into(), // default name
            ]),
            actions_events: Some(vec![
                KeyCode::Char('f').into(), // name: 'foo'
                KeyCode::Char('o').into(), //
                KeyCode::Char('o').into(), //
                KeyCode::Enter.into(),     //
            ]),
        }),
    )
    .unwrap());
    assert_yaml_snapshot!(fs::read_to_string("tests-out/my-project1/test.txt").unwrap());
}

#[test]
#[serial]
fn test_run_with_local_project_actions_git_mode() {
    // dont run this in CI, git requires a registered identity
    if env::var("CI").is_err() {
        let res = run_with_no_config(
            Some("rusty-ferris-club/backpack-e2e-frozen-localproj"),
            None,
            CopyMode::Copy,
            true,
            Some(RunnerEvents {
                prompt_events: Some(vec![
                    KeyCode::Enter.into(), // default name
                ]),
                actions_events: Some(vec![
                    KeyCode::Char('f').into(), // name: 'foo'
                    KeyCode::Char('o').into(), //
                    KeyCode::Char('o').into(), //
                    KeyCode::Enter.into(),     //
                ]),
            }),
        )
        .unwrap();
        assert_yaml_snapshot!(res);
        assert_yaml_snapshot!(fs::read_to_string("tests-out/my-project1/test.txt").unwrap());
    }
}

#[test]
#[serial]
fn test_run_with_actions() {
    assert_debug_snapshot!(run_with_local_config(
        Some("integration"),
        None,
        CopyMode::Apply,
        r#"
projects:
  integration:
    shortlink: rusty-ferris-club/backpack-e2e-frozen
    actions:
    - name: "test confirm"
      interaction:
        kind: confirm
        prompt: "are you sure?"
    - name: "grab input and create file"
      interaction:
        kind: input
        prompt: name of your thing
        out: name
      run: touch {{name}}.txt
"#,
        false,
        Some(RunnerEvents {
            actions_events: Some(vec![
                KeyCode::Char('y').into(), // yes
                KeyCode::Enter.into(),     //
                KeyCode::Char('t').into(), // city: 'tlv'
                KeyCode::Char('l').into(), //
                KeyCode::Char('v').into(), //
                KeyCode::Enter.into(),     //
            ]),
            prompt_events: Some(vec![
                KeyCode::Enter.into(),     // no dest
            ]),
        }),
    ));
}

#[test]
#[serial]
fn test_run_with_actions_say_no() {
    assert_debug_snapshot!(run_with_local_config(
        Some("integration"),
        None,
        CopyMode::Apply,
        r#"
projects:
  integration:
    shortlink: rusty-ferris-club/backpack-e2e-frozen
    actions:
    - name: "test confirm"
      interaction:
        kind: confirm
        prompt: "are you sure?"
      break_if_cancel: true
    - name: "grab input and create file"
      interaction:
        kind: input
        prompt: name of your thing
        out: name
      run: touch {{name}}.txt
"#,
        false,
        Some(RunnerEvents {
            actions_events: Some(vec![
                KeyCode::Char('n').into(), // yes
                KeyCode::Enter.into(),     //
            ]),
            prompt_events: Some(vec![
                KeyCode::Enter.into(),     // no dest
            ]),
        }),
    ));
}

#[test]
#[serial]
fn test_run_with_actions_say_yes() {
    assert_debug_snapshot!(run_with_local_config(
        Some("integration"),
        None,
        CopyMode::Apply,
        r#"
projects:
  integration:
    shortlink: rusty-ferris-club/backpack-e2e-frozen
    actions:
    - name: "test confirm"
      interaction:
        kind: confirm
        prompt: "are you sure?"
      break_if_cancel: true
    - name: "grab input and create file"
      interaction:
        kind: input
        prompt: name of your thing
        out: name
      run: touch {{name}}.txt
"#,
        false,
        Some(RunnerEvents {
            actions_events: Some(vec![
                KeyCode::Char('y').into(), // yes
                KeyCode::Enter.into(),     //
                KeyCode::Char('t').into(), // 'tlv'
                KeyCode::Char('l').into(), //
                KeyCode::Char('v').into(), //
                KeyCode::Enter.into(),     //
            ]),
            prompt_events: Some(vec![
                KeyCode::Enter.into(),     // no dest
            ]),
        }),
    ));
}

#[test]
#[serial]
fn test_run_with_actions_say_no_without_break() {
    assert_debug_snapshot!(run_with_local_config(
        Some("integration"),
        None,
        CopyMode::Apply,
        r#"
projects:
  integration:
    shortlink: rusty-ferris-club/backpack-e2e-frozen
    actions:
    - name: "test confirm"
      interaction:
        kind: confirm
        prompt: "are you sure?"
    - name: "grab input and create file"
      interaction:
        kind: input
        prompt: name of your thing
        out: name
      run: touch {{name}}.txt
"#,
        false,
        Some(RunnerEvents {
            actions_events: Some(vec![
                KeyCode::Char('n').into(), // yes
                KeyCode::Enter.into(),     //
                KeyCode::Char('t').into(), // 'tlv'
                KeyCode::Char('l').into(), //
                KeyCode::Char('v').into(), //
                KeyCode::Enter.into(),     //
            ]),
            prompt_events: Some(vec![
                KeyCode::Enter.into(),     // no dest
            ]),
        }),
    ));
}
