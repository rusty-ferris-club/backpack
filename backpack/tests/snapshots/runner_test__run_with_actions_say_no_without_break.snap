---
source: backpack/tests/runner_test.rs
expression: "run_with_local_config(Some(\"integration\"), None, CopyMode::Apply,\n    r#\"\nprojects:\n  integration:\n    shortlink: rusty-ferris-club/backpack-e2e-frozen\n    actions:\n    - name: \"test confirm\"\n      interaction:\n        kind: confirm\n        prompt: \"are you sure?\"\n    - name: \"grab input and create file\"\n      interaction:\n        kind: input\n        prompt: name of your thing\n        out: name\n      run: touch {{name}}.txt\n\"#,\n    false,\n    Some(RunnerEvents {\n            actions_events: Some(vec![KeyCode :: Char('n').into(), KeyCode ::\n                    Enter.into(), KeyCode :: Char('t').into(), KeyCode ::\n                    Char('l').into(), KeyCode :: Char('v').into(), KeyCode ::\n                    Enter.into(),]),\n            prompt_events: Some(vec![KeyCode :: Enter.into(),]),\n        }))"
---
Ok(
    [
        "tests-out/.backpack.yaml",
        "tests-out/.cargo/config.toml",
        "tests-out/.github/ISSUE_TEMPLATE/bug_report.yml",
        "tests-out/.github/ISSUE_TEMPLATE/feature_request.yml",
        "tests-out/.github/workflows/build.yml",
        "tests-out/.github/workflows/release.yml",
        "tests-out/.gitignore",
        "tests-out/Cargo.lock",
        "tests-out/Cargo.toml",
        "tests-out/LICENSE.txt",
        "tests-out/README.md",
        "tests-out/bumblefoot/Cargo.toml",
        "tests-out/bumblefoot/build.rs",
        "tests-out/bumblefoot/src/bin/bumblefoot.rs",
        "tests-out/bumblefoot/src/bin/cmd/default.rs",
        "tests-out/bumblefoot/src/bin/cmd/mod.rs",
        "tests-out/bumblefoot/src/bin/cmd/validate.rs",
        "tests-out/bumblefoot/src/data.rs",
        "tests-out/bumblefoot/src/lib.rs",
        "tests-out/bumblefoot/src/runner.rs",
        "tests-out/media/cover.png",
        "tests-out/media/screen.png",
        "tests-out/tlv.txt",
        "tests-out/xtask/Cargo.toml",
        "tests-out/xtask/src/dual/Cargo.toml",
        "tests-out/xtask/src/dual/src/bin/bumblefoot.rs",
        "tests-out/xtask/src/dual/src/bin/cmd/default.rs",
        "tests-out/xtask/src/dual/src/bin/cmd/mod.rs",
        "tests-out/xtask/src/dual/src/bin/cmd/validate.rs",
        "tests-out/xtask/src/main.rs",
        "tests-out/xtask/src/simple/Cargo.toml",
        "tests-out/xtask/src/simple/src/main.rs",
    ],
)
